crates/storage/nippy-jar/src/compression/zstd.rs

```diff
@@ -0,0 +1,236 @@
+                tmp_buf.reserve(column_value.len() * multiplier);
+                multiplier += 1;
+                if multiplier == 5 {
+                    return Err(NippyJarError::Disconnect(err))
+                }
+            }
+
+            handle.write_all(tmp_buf)?;
+            tmp_buf.clear();
+        } else {
+            handle.write_all(column_value)?;
+        }
+
+        Ok(())
+    }
+
+    /// Decompresses a value using a dictionary to a user provided buffer.
+    pub fn decompress_with_dictionary(
+        column_value: &[u8],
+        output: &mut Vec<u8>,
+        decompressor: &mut Decompressor<'_>,
+    ) -> Result<(), NippyJarError> {
+        let mut multiplier = 1;
+
+        // Just an estimation.
+        let required_capacity = column_value.len() * 2;
+
+        output.reserve(required_capacity.saturating_sub(output.capacity()));
+
+        // Decompressor requires the destination buffer to be big enough to write to, otherwise it
+        // fails. However, we don't know how big it will be. We keep retrying.
+        // If we eventually fail, it probably means it's another kind of error.
+        while let Err(err) = decompressor.decompress_to_buffer(column_value, output) {
+            output.reserve(
+                Decompressor::upper_bound(column_value).unwrap_or(required_capacity) * multiplier,
+            );
+
+            multiplier += 1;
+            if multiplier == 5 {
+                return Err(NippyJarError::Disconnect(err))
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl Compression for Zstd {
+    fn decompress(&self, value: &[u8]) -> Result<Vec<u8>, NippyJarError> {
+        let mut decompressed = Vec::with_capacity(value.len() * 2);
+        let mut decoder = zstd::Decoder::new(value)?;
+        decoder.read_to_end(&mut decompressed)?;
+        Ok(decompressed)
+    }
+
+    fn compress_to<W: Write>(&self, src: &[u8], dest: &mut W) -> Result<(), NippyJarError> {
+        let mut encoder = zstd::Encoder::new(dest, self.level)?;
+        encoder.write_all(src)?;
+
+        encoder.finish()?;
+
+        Ok(())
+    }
+
+    fn compress(&self, src: &[u8]) -> Result<Vec<u8>, NippyJarError> {
+        let mut compressed = Vec::with_capacity(src.len());
+
+        self.compress_to(src, &mut compressed)?;
+
+        Ok(compressed)
+    }
+
+    fn is_ready(&self) -> bool {
+        matches!(self.state, ZstdState::Ready)
+    }
+
+    /// If using it with dictionaries, prepares a dictionary for each column.
+    fn prepare_compression(
+        &mut self,
+        columns: Vec<impl IntoIterator<Item = Vec<u8>>>,
+    ) -> Result<(), NippyJarError> {
+        if !self.use_dict {
+            return Ok(())
+        }
+
+        // There's a per 2GB hard limit on each column data training, iterator should take care of
+        // it. Each element should ideally be bigger than 1 (otherwise sigev) huffman coding
+        // + sueprstring ?
```

Comment:
can you expand on it a bit? I think I don't fully understand what does it mean

Reply:
```
 // There's a per 2GB hard limit on each column data set for training
        // REFERENCE: https://github.com/facebook/zstd/blob/dev/programs/zstd.1.md#dictionary-builder
        // ```
        // -M#, --memory=#: Limit the amount of sample data loaded for training (default: 2 GB).
        // Note that the default (2 GB) is also the maximum. This parameter can be useful in
        // situations where the training set size is not well controlled and could be potentially
        // very large. Since speed of the training process is directly correlated to the size of the
        // training sample set, a smaller sample set leads to faster training.`
        // ```
   ```
     
cleaned it up but will put it here as well.